### PR TITLE
Deepblast

### PR DESCRIPTION
**Pull request by mortonjt**
This updates the deepblast aligner.  This code is not tested yet, so don't merge.

We probably want to rename this script to something like `tmvec-search` instead of `tm_vec_run.py` so that it is more informative.

### CODE REVIEW

**Comment by mortonjt** [comment]
I'm also noticing that the refactored functions like `encode` haven't made it into the CLI scripts yet.  We'll need to add those in.

File: scripts/tm_vec_run.py [hunk 1]
```diff
@@ -168,12 +197,15 @@
...
+if args.deepblast_model is not None:
+    align_model = DeepBLAST.load_from_checkpoint(
+        args.deepblast_model, lm=model, tokenizer=tokenizer,
+        alignment_mode=args.alignment_mode,  # TODO
+        device=args.device)
     align_model = align_model.to(device)
 
     #Read in database sequences
-    with open(args.database_sequences) as handle_db:
+    with open(args.database) as handle_db:
```
Comment (mortonjt) [inline scripts/tm_vec_run.py:208]: @tymor22 I'm getting errors here.  I'm a little confused here. Do we really need to loop through all of the sequences in the database?

Comment (mortonjt) [reply scripts/tm_vec_run.py:208]: If we are really passing in the database sequences separately, we also need to make all of those fasta files available as well.

Comment (mortonjt) [reply scripts/tm_vec_run.py]: This has been resolved

**Comment by mortonjt** [comment]
Hi @tymor22 this pull request is now operational.  Could you take a look at it and merge it when you have the chance?

File: example/2-search-database.sh [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+python /mnt/home/jmorton/ceph/research/gert/deep_blast_training/collect_env_details.py
+
+
+#wget https://users.flatironinstitute.org/thamamsy/public_www/embeddings_cath_s100_final.npy
+#wget https://users.flatironinstitute.org/thamamsy/public_www/embeddings_cath_s100_w_metadata.tsv
+#wget https://users.flatironinstitute.org/thamamsy/public_www/cath_large.npy
+#wget https://users.flatironinstitute.org/thamamsy/public_www/cath_large_metadata.npy
+#wget https://users.flatironinstitute.org/thamamsy/public_www/cath-domain-seqs-large.fa
+#wget https://users.flatironinstitute.org/thamamsy/public_www/cath-domain-seqs-large.fai
+#wget https://users.flatironinstitute.org/jmorton/public_www/deepblast-public-data/checkpoints/deepblast-l8.ckpt
```
Comment (mortonjt) [inline example/2-search-database.sh:26]: Need to be uncommented

File: example/2-search-database.sh [hunk 3]
```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
+#SBATCH --mail-type=ALL         # Mail events (NONE, BEGIN, END, FAIL, ALL)
+#SBATCH --mail-user=jmorton@flatironinstitute.org
```
Comment (mortonjt) [inline example/2-search-database.sh:3]:
```suggestion
```

File: example/2-search-database.sh [hunk 4]
```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
+#SBATCH --mail-type=ALL         # Mail events (NONE, BEGIN, END, FAIL, ALL)
```
Comment (mortonjt) [inline example/2-search-database.sh:2]:
```suggestion
```

File: example/2-search-database.sh [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+#SBATCH --tasks-per-node=1
+
+source ~/ceph/venv/deepblast/bin/activate
+
+module -q purge
+module load gcc python cuda cudnn
+
+export CUDA_HOME=$CUDA_BASE
+
+python /mnt/home/jmorton/ceph/research/gert/deep_blast_training/collect_env_details.py
```
Comment (mortonjt) [inline example/2-search-database.sh:17]:
```suggestion
```

File: example/2-search-database.sh [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+
+#wget https://users.flatironinstitute.org/thamamsy/public_www/embeddings_cath_s100_final.npy
```
Comment (mortonjt) [inline example/2-search-database.sh:20]:
```suggestion
wget https://users.flatironinstitute.org/thamamsy/public_www/embeddings_cath_s100_final.npy
```

File: example/2-search-database.sh [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+#wget https://users.flatironinstitute.org/thamamsy/public_www/embeddings_cath_s100_w_metadata.tsv
```
Comment (mortonjt) [inline example/2-search-database.sh:21]:
```suggestion
wget https://users.flatironinstitute.org/thamamsy/public_www/embeddings_cath_s100_w_metadata.tsv
```

File: example/2-search-database.sh [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+#wget https://users.flatironinstitute.org/thamamsy/public_www/cath_large.npy
```
Comment (mortonjt) [inline example/2-search-database.sh:22]:
```suggestion
wget https://users.flatironinstitute.org/thamamsy/public_www/cath_large.npy
```

File: example/2-search-database.sh [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+#wget https://users.flatironinstitute.org/thamamsy/public_www/cath_large_metadata.npy
+#wget https://users.flatironinstitute.org/thamamsy/public_www/cath-domain-seqs-large.fa
```
Comment (mortonjt) [inline example/2-search-database.sh:24]:
```suggestion
wget https://users.flatironinstitute.org/thamamsy/public_www/cath-domain-seqs-large.fa
```

File: example/2-search-database.sh [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+#wget https://users.flatironinstitute.org/thamamsy/public_www/cath-domain-seqs-large.fai
```
Comment (mortonjt) [inline example/2-search-database.sh:25]:
```suggestion
wget https://users.flatironinstitute.org/thamamsy/public_www/cath-domain-seqs-large.fai
```

File: example/2-search-database.sh (hunk 2 above)
Comment (mortonjt) [inline example/2-search-database.sh:26]:
```suggestion
wget https://users.flatironinstitute.org/jmorton/public_www/deepblast-public-data/checkpoints/deepblast-l8.ckpt
```

File: scripts/tm_vec_run.py [hunk 11]
```diff
@@ -30,44 +31,73 @@
...
 
+parser.add_argument("--database-fasta",
+        type=Path,
+        required=True,
+        help="Database that contains the corresponding protein sequences in fasta format."
+)
+
+parser.add_argument("--database-faidx",
+        type=Path,
+        required=True,
```
Comment (mortonjt) [inline scripts/tm_vec_run.py:42]:
```suggestion
        required=False,
```

File: scripts/tm_vec_run.py [hunk 12]
```diff
@@ -30,44 +31,73 @@
         help="Database to query"
 )
 
+parser.add_argument("--database-fasta",
+        type=Path,
+        required=True,
```
Comment (mortonjt) [inline scripts/tm_vec_run.py:36]:
```suggestion
        required=False,
```

File: scripts/tm_vec_run.py [hunk 13]
```diff
@@ -150,43 +195,46 @@
...
-            pred_alignment = align_model.align(x, y)
-            x_aligned, y_aligned = states2alignment(pred_alignment, x, y)
-            alignments_i.append([x_aligned, pred_alignment, y_aligned])
+            seq_i = metadata_database[I[i, j]]
+
+            y = seq_db[seq_i]
+            try :
+                pred_alignment = align_model.align(x, y)
+                # Note : there is an edge case that smith-waterman will throw errors, but
+                # needleman-wunsch won't.
```
Comment (mortonjt) [inline scripts/tm_vec_run.py:233]:
```suggestion
```

File: scripts/tm_vec_run.py [hunk 14]
```diff
@@ -150,43 +195,46 @@
...
-            y = seqs_db[I[i,j]]
-            pred_alignment = align_model.align(x, y)
-            x_aligned, y_aligned = states2alignment(pred_alignment, x, y)
-            alignments_i.append([x_aligned, pred_alignment, y_aligned])
+            seq_i = metadata_database[I[i, j]]
+
+            y = seq_db[seq_i]
+            try :
+                pred_alignment = align_model.align(x, y)
+                # Note : there is an edge case that smith-waterman will throw errors, but
```
Comment (mortonjt) [inline scripts/tm_vec_run.py:232]:
```suggestion
                # Note : there is an edge case that the aligner will throw errors
```

File: example/2-search-database.sh [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+#wget https://users.flatironinstitute.org/jmorton/public_www/deepblast-public-data/checkpoints/deepblast-l8.ckpt
+
+
+tm_vec_run.py \
+    --query test.fa \
```
Comment (mortonjt) [inline example/2-search-database.sh:30]:
```suggestion
    --query bagel.fa \
```

File: scripts/tm_vec_run.py [hunk 16]
```diff
@@ -150,43 +195,46 @@
...
+        alignment_mode=args.alignment_mode,  # TODO
+        device=device)
     align_model = align_model.to(device)
-
-    #Read in database sequences
-    with open(args.database_sequences) as handle_db:
-        seqs_db = []
-        for record in SeqIO.parse(handle_db, "fasta"):
-            seqs_db.append(str(record.seq))
+    seq_db = Indexer(args.database_fasta, args.database_faidx)
```
Comment (mortonjt) [inline scripts/tm_vec_run.py:219]: TODO : throw error if these aren't available.

**Review by tymor22 — APPROVED** [review]

File: example/1-build-database.sh [hunk 17]
```diff
@@ -0,0 +1,9 @@
+#wget https://users.flatironinstitute.org/thamamsy/public_www/tm_vec_cath_model.ckpt
+#wget https://users.flatironinstitute.org/thamamsy/public_www/tm_vec_cath_model_params.json
```
Comment (mortonjt) [inline example/1-build-database.sh:2]: TODO need to fix

File: example/2-search-database.sh [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+export CUDA_HOME=$CUDA_BASE
+
+
+
+wget https://users.flatironinstitute.org/thamamsy/public_www/embeddings_cath_s100_final.npy
+wget https://users.flatironinstitute.org/thamamsy/public_www/embeddings_cath_s100_w_metadata.tsv
+wget https://users.flatironinstitute.org/thamamsy/public_www/cath_large.npy
+wget https://users.flatironinstitute.org/thamamsy/public_www/cath_large_metadata.npy
+wget https://users.flatironinstitute.org/thamamsy/public_www/cath-domain-seqs-large.fa
+#wget https://users.flatironinstitute.org/thamamsy/public_www/cath-domain-seqs-large.fai
```
Comment (mortonjt) [inline example/2-search-database.sh:22]: TODO fix

**Pull request merged**